### PR TITLE
Task 6 Perceiver Model to classify AD / NC brain scans from ADNI Dataset

### PR DESCRIPTION
**Pull request by dcpais**
This project implements the Perceiver transformer as requested in task 6 of the report. The dataset that the perceiver classifies is the ADNI AD / NC brain scans. 

The model achieved an accuracy of 50% which is not very meaningful. The model itself is very closely based on a formulated model from the following paper: https://arxiv.org/abs/2103.03206. Within my own model, there are slight differences in architecture (mainly to do with the cross-attention structure used). However, I have chosen to adhere to similar hyperparameters that they use in experiments within the paper, as they have found success with them.

The key difference in my model is the lack of integrated positional embedding. There were problems setting up positional embedding, and as such it was omitted and replaced with a simple linear layer. 

The pull request contains 5 files:
1. train.py: the script used to train the model
2. modules.py: the script that contained the pytorch modules used in the perceiver architecture
3. predict.py: the script used to test the trained model and predict on the test set
4. dataset.py: the script used to load data for the model to train / test on
5. README.md: A detailed report on the model. It outlines the model architecture, how it was made as well as its results.

### CODE REVIEW

**Comment by LinfengLiu98** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme:
Good
Need more explanation on how the perceiver works.

Commit messages: Excellent, detailed.

Code:
- Uses Perceiver transformer
- Well commented
- Good model design

Functionality/Performance:
- Test ACC: 50.44% 
- Preprocssing data correctly
- Model has bugs inside and not learning
- No patient level split.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem, performance low < 0.6 -1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable, missing usage -1
Model/technical explanation could be better -1
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Feedback required, remove pyc files -2
Request Description good

**Comment by shakes76** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks granted.
